src/instructlab/train/linux_train.py

```diff
@@ -173,6 +175,52 @@ def linux_train(
         hpu.init()
         report_hpu_device(device)
 
+    # device register a module, e.g. torch.cpu or torch.cuda
+    device_module = getattr(torch, device.type, None)
+    # bfloat16 is not supported on older CUDA versions and devices
+    # with CUDA support level < 8.0.
+    if hasattr(device_module, "is_bf16_supported"):
+        use_bf16 = device_module.is_bf16_supported()
+        use_fp16 = not use_bf16
+    elif device.type == "cpu":
+        # TODO: check if Torch and CPU support AVX2, F16C, AVX512
+        use_bf16 = False
+        use_fp16 = False
+    else:
+        # assume bf16 supported unless device says otherwise
+        use_bf16 = True
+        use_fp16 = False
+
+    torch_dtype = "auto" if device.type == "cuda" else None
```

Comment:
I think it would be super useful to document why torch_dtype=None is faster on CPUs 

So on CPUs, `dtype=float32` will be faster than `dtype=bfloat16`. Why? A link to a nice explanation of that would be great

I guess `torch_dtype=None` gives a `dtype=float32` on CPUs? Why? I can't quickly find any docs that explains that - does None just mean that we use whatever torch.get_default_dtype() gives, which is float32 by default? 

What dtype does `torch_dtype=auto` give on CPUs? Sounds like `dtype=bfloat16`? Why? It's detecting that this particular model was saved with that dtype?

Can't seem to find the docs on hugginface.co/docs so see: https://github.com/huggingface/transformers/blob/f3f640dce14bee3b3930a774c3dfac92977eee7f/src/transformers/modeling_utils.py#L2878-L2898

We don't have a dtype in the model's [config.json](https://huggingface.co/instructlab/merlinite-7b-lab-GGUF/blob/main/config.json) so that doesn't seem to be a factor, but it could be

If the torch_dtype=auto behavior is model-specific, but we know we want float32 except on low-memory systems ... maybe for CPUs, we should just explicitly set torch_dtype to either float32 or bfloat16?

Reply:
It very much depends on the hardware and compiler. In general, x86_64 CPUs have support for standard precision and double precision floats (fp32, fp64). Half precision instructions (fp16) were added in ISA level x86_64-v3 and brain float (bf16) SIMD instruction were added in x86_64-v4. The document https://pytorch.org/docs/stable/amp.html#ops-that-can-autocast-to-float16 explains autocasting.

AFAIK we want to use

* *bf16* when the CPU or hardware accelerator supports it
* fall back to *fp16* on Nvidia hardware with CUDA compute level < 8.0
* fall back to *fp32* on CPUs

Reply:
Based on @tiran 's comment, it's possible that my laptop has this support while my server does not:

> * **bf16** when the CPU or hardware accelerator supports it

which could explain why `"auto"` gets the best performance on my laptop.

If `"auto"` isn't doing an adequate job of checking if that is supported and is choosing it even if it's not actually supported on my server, maybe that's killing the performance. I'll have to keep digging here ...

Reply:
What is the output of `torch.backends.cpu.get_cpu_capability()` on your server and your laptop?

Reply:
On my laptop -- 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz

```
>>> torch.backends.cpu.get_cpu_capability()
'AVX512'
```

On the server -- AMD EPYC 7R32

```
>>> torch.backends.cpu.get_cpu_capability()
'AVX2'
```